lib/statsd/instrument/udp_sink.rb

```diff
@@ -5,45 +5,72 @@ module Instrument
     # @note This class is part of the new Client implementation that is intended
     #   to become the new default in the next major release of this library.
     class UDPSink
-      def self.for_addr(addr)
+      def self.for_addr(addr, **kwargs)
         host, port_as_string = addr.split(":", 2)
-        new(host, Integer(port_as_string))
+        new(host, Integer(port_as_string), **kwargs)
       end
 
       attr_reader :host, :port
 
-      def initialize(host, port)
+      class << self
+        def finalizer(sink)
+          proc { sink.flush }
+        end
+      end
+
+      def initialize(host, port, batch_size: 1)
         @host = host
         @port = port
         @mutex = Mutex.new
         @socket = nil
+
+        require "concurrent/array"
+        @buffer = Concurrent::Array.new
+        @batch_size = batch_size
       end
 
       def sample?(sample_rate)
         sample_rate == 1 || rand < sample_rate
       end
 
       def <<(datagram)
-        with_socket { |socket| socket.send(datagram, 0) }
+        @buffer << datagram
+        if @buffer.size >= @batch_size
+          flush
+        end
         self
+      end
+
+      NEWLINE = "\n".b.freeze
+      def flush
+        return if @buffer.empty?
+
+        synchronize do
+          return if @buffer.empty?
+          messages = @buffer.shift(@buffer.size)
+          emit_datagram(messages.join(NEWLINE))
```

Comment:
While I know that Datadog will deal with newlines properly, I am not sure if this is the case for all StatsD server implementations. But maybe we don't have to care about that.

Reply:
Thanks for the info. I'm actually not that knowledgeable about statsd outside of Datadog.

I can gate all this behavior under the `datadog` implementation.

Reply:
The original `statsd` implementation also deals with newlines correctly. I am OK with assuming this always works - we don'r really support other implementations anyway.